src/main/groovy/com/github/winplay02/gitcraft/MinecraftVersionGraph.java

```diff
@@ -105,39 +103,118 @@ private void testGraphConnectivity() {
 			}
 		}
 		long amountRootNodes = this.edgesBack.entrySet().stream().filter(entry -> entry.getValue().isEmpty()).count();
-		if (amountRootNodes != 1) {
-			List<OrderedVersion> rootVersions = this.edgesBack.entrySet().stream().filter(entry -> entry.getValue().isEmpty()).map(Map.Entry::getKey).toList();
-			if (rootVersions.size() == 2 && rootVersions.get(0).hasClientCode() != rootVersions.get(1).hasClientCode() && rootVersions.get(0).hasServerCode() != rootVersions.get(1).hasServerCode()) {
-				return;
+		if (amountRootNodes < 1) {
+			MiscHelper.panic("There is no root node. This either means, that the version graph is empty, or that it contains a cycle.");
+		}
+	}
+
+	private void findMainBranch() {
+		this.roots.clear();
+		this.branchPoints.clear();
+
+		Set<OrderedVersion> roots = this.edgesBack.entrySet()
+			.stream()
+			.filter(entry -> entry.getValue().isEmpty())
+			.map(Map.Entry::getKey)
+			.collect(Collectors.toSet());
+
+		for (OrderedVersion root : roots) {
+			int length = this.findBranchPoints(root);
+			this.roots.put(root, length);
+		}
+		for (OrderedVersion root : roots) {
+			this.markMainBranch(root);
+		}
+		this.getDeepestRootVersion();
+	}
+
+	private int findBranchPoints(OrderedVersion mcVersion) {
+		int branchLength = this.branchPoints.getOrDefault(mcVersion, 0);
+
+		// check if this branch point was already identified earlier
+		if (branchLength > 0) {
+			return branchLength;
+		}
+
+		NavigableSet<OrderedVersion> nextBranches = this.getFollowingNodes(mcVersion);
+
+		// tip of a branch
+		if (nextBranches.size() < 1) {
+			branchLength = 0;
+		}
+
+		// only one following node; no splitting
+		if (nextBranches.size() == 1) {
+			OrderedVersion nextBranch = nextBranches.first();
+			int nextBranchLength = this.findBranchPoints(nextBranch);
+
+			branchLength = nextBranchLength;
+		}
+
+		// multiple following nodes; find largest path forward
+		if (nextBranches.size() > 1) {
+			for (OrderedVersion nextBranch : this.getFollowingNodes(mcVersion)) {
+				// since the branch splits, mark these versions as branch points
+				int nextBranchLength = this.findBranchPoints(nextBranch);
+				this.branchPoints.put(nextBranch, nextBranchLength);
+
+				if (nextBranchLength > branchLength) {
+					branchLength = nextBranchLength;
+				}
 			}
-			MiscHelper.panic(amountRootNodes < 1 ? "There is no root node. This either means, that the version graph is empty, or that it contains a cycle." : "There are multiple root nodes. A connected git history would not be guaranteed");
 		}
+
+		return branchLength + 1;
 	}
 
-	protected static final Pattern LINEAR_SNAPSHOT_REGEX = Pattern.compile("(^\\d\\dw\\d\\d[a-z]$)|(^\\d.\\d+(.\\d+)?(-(pre|rc)\\d+|_[a-z_\\-]+snapshot-\\d+| Pre-Release \\d+)?$)");
+	private void markMainBranch(OrderedVersion mcVersion) {
+		// after the branch points have been identified,  search for the longest path
```

Comment:
That is a good point, that I hadn't considered. I wasn't fond of the version-id-check approach for a few reasons. It enforces the same branch structure for any manifest, while different manifests might order versions differently. It also assumes a certain structure in the version ids. Both of these were causing issues with Skyrisings manifest. I am not sure what the best approach is though. Keeping a hardcoded list of version ids for each manifest won't do, as you might have to update that list when a new version drops. Perhaps the pattern matching could be re-implemented in some way, but with different patterns for each manifest. Still feels like a hotfix, but it would solve the issue at least.

Reply:
I added the id pattern matching back, with slightly different patterns for both manifests.

Reply:
Moving the matching logic to the metadata provider seems like the best way to go forward, for now